Revitalise CONTRIBUTING.md (#3907)
PaperSpigot/Paper: CONTRIBUTING.md

Contributing to Paper
==========================

> PaperMC is very happy you're willing to contribute to the projects we have, and
> we are very lentient with the PRs provided, but there are still some guidelines
> you should follow.

Reviewer: ```suggestion
PaperMC is happy you're willing to contribute to our projects. We are usually
very lenient with all submitted PRs, but there are still some guidelines you
can follow to make the approval process go more smoothly.
```

* [Adding Patches](#adding-patches)
* [Modifying Patches](#modifying-patches)
  * [Method 1](#method-1)
    * [Using the Paper tool](#using-the-paper-tool)
    * [Manual method: Stashing](#manual-method-stashing)
  * [Method 2 - Fixup commits](#method-2---fixup-commits)
    * [Manual method](#manual-method)
    * [Automatic method](#automatic-method)
* [PR Policy](#pr-policy)
* [Formatting](#formatting)
* [Patch Notes](#patch-notes)
* [Obfuscation Helpers](#obfuscation-helpers)
* [Configuration files](#configuration-files)
  * [PaperConfig example](#paperconfig-example)
  * [PaperWorldConfig example](#paperworldconfig-example)
* [Frequently Asked Questions](#frequently-asked-questions)
  * [I can't find the NMS file I need!](#i-cant-find-the-nms-file-i-need)
  * [Where can I learn what this method/field does?](#where-can-i-learn-what-this-methodfield-does)
  * [My commit doesn't need a build, what do I do?](#my-commit-doesnt-need-a-build-what-do-i-do)
  * [Patching and building is *really* slow, what can I do?](#patching-and-building-is-really-slow-what-can-i-do)

<!-- vim-markdown-toc -->

## Use a Personal Fork and not Organization

> Paper will routinely modify your PR for you to rebase it on latest or take care
> of any minor nitpicks that we can just solve for you instead of making you go
> back for a trivial item.

Reviewer: ```suggestion
Paper will routinely modify your PR; Whether it's just a quick rebase or to take care
of any minor nitpicks we might have. Often times it's better for us to solve these
problems for you than to make you go back and forth trying to fix it yourself.
```

> However, if you use an Organization for the PR, this prevents Paper from doing

Reviewer 2: Organization -> organization

Author: The original document had capital O, so I have a feeling that's on purpose.

Reviewer 2: Alright, I see why that was done.

> However, if you use an Organization for the PR, this prevents Paper from doing
> this, and will require us to manually merge it and close the PR instead of it
> being marked as the purple merged.

Reviewer: ```suggestion
Unfortunately, if you use an organization for your PR, it prevents Paper from modifying
it. This requires us to manually merge your PR; Resulting in us closing the PR itself
instead of it being marked as merged.
```

> We much prefer to have merged PR's show as merged, so please do not use
> organizations repositories for PR as GitHub sadly does not support maintainers
> pushing to them.

Reviewer: ```suggestion
We much prefer to have PRs show as merged, so please do not use organizations
repositories for PRs.

See <https://github.com/isaacs/github/issues/1681> for more information on the
issue.
```

> See <https://github.com/isaacs/github/issues/1681> for more information on the
> issue.

Reviewer: ```suggestion
```

Author: The link was present in the original document, and provides reasoning for the requirement. If someone sees it when it's closed, we can also remove this section from the file altogether.

## Requirements

> To actually start doing the more enjoyable part of "PR or shut up", you have
> to get some duct tape.

Reviewer: I'm all for jokes but the newbies may not get this :P
```suggestion
To actually get started with your pull request, you're gonna need some duct tape.
```

> To do the other part, PRing changes, you require the following software, which

Reviewer 3: ```suggestion
To get started with PRing changes, you require the following software, which
```

What other part?

> To do the other part, PRing changes, you require the following software, which
> are all available in (most) package managers such as `apt`, `pacman`, `xbps`,
> `homebrew` (macOS / Linux), and more:

Reviewer: ```suggestion
You'll need the following software, most of which
are all available in (most) package managers such as `apt`, `pacman`, `xbps`,
`homebrew` (macOS / Linux), and more:
```

- `git` (package `git` everywhere);
- `patch` (often package `patch`);
- A Java 8 or later JDK (packages vary, use Google/DuckDuckGo/etc.).  
If you need one, you can find them on [AdoptOpenJDK](https://adoptopenjdk.net/).
- `maven` (often package `maven`; can be found on
[Apache's site](https://maven.apache.org/download.cgi) too).

If you're compiling with Docker, you can use the
[`adoptopenjdk`](https://hub.docker.com/_/adoptopenjdk/) images like so:

```console
# docker run -it -v "$(pwd)":/data --rm adoptopenjdk:8-jdk-hotspot bash
Pulling image...

root@abcdefg1234:/# javac -version
javac 1.8.0_252
```

## Understanding Patches

> Paper is made to be built on top of patches applied on top of Spigot in what are
> essentially submodules within the repository. These are centered around:

Reviewer: ```suggestion
Paper is, essentially, a set of patches applied on top of Spigot. These patches are
split into different folders which target a certain part of code;
```

> - `Paper-API` (the `Spigot-API`/`Bukkit` for Paper);
> - `Paper-MojangAPI` (API for
> [Mojang's Brigadier](https://github.com/Mojang/brigadier));
> - `Paper-Server` (the `Spigot`/`CraftBukkit` for Paper).

Reviewer: ```suggestion
- `Paper-API` - Modifications to the `Spigot-API`/`Bukkit`;
- `Paper-MojangAPI` - An API for
[Mojang's Brigadier](https://github.com/Mojang/brigadier);
- `Paper-Server` - Modifications to `Spigot`/`CraftBukkit`.
```

of how to use git is required. A basic tutorial can be found here:
<https://git-scm.com/docs/gittutorial>.

Assuming you already have forked the repository:

1. Pull the latest changes from the main repository;
1. Type `./paper patch` in git bash to apply the changes from upstream;
1. cd into `Paper-Server` for server changes, and `Paper-API` for API changes.  
You can also run `./paper server` or `./paper api` for these same directories
respectively. If you install the Paper tool with `./paper setup`, you can run
`paper <command>` from anywhere in most cases.

These directories aren't git repositories in the traditional sense:

- Every single commit in `Paper-Server`/`Paper-API` is a patch;
- `origin/master` points to a directory similar to `Paper-Server`/`Paper-API`
but for Paper;
- Typing `git status` should show that we are 10 or 11 commits ahead of master,
meaning we have 10 or 11 patches Spigot and CraftBukkit don't.
   - If it says something like `212 commits ahead, 207 commits behind`,
   cd into the root directory of the cloned repository and type `git fetch` to
   update your upstream. Setting up a remote for the upstream Paper repository

>    might be necessary.

Reviewer: ```suggestion
Assuming you already have forked the repository:

1. Clone your fork to your local machine
1. Type `./paper patch` in git bash to apply the changes from upstream;
1. cd into `Paper-Server` for server changes, and `Paper-API` for API changes.  
You can also run `./paper server` or `./paper api` for these same directories
respectively. 
1. You can also run `./paper setup`, which allows you to only have to type
`paper <command>` from anywhere in most cases.

`Paper-Server` and `Paper-API` aren't git repositories in the traditional sense:

- Every single commit in `Paper-Server`/`Paper-API` is a patch;
- `origin/master` points to a directory similar to `Paper-Server`/`Paper-API`
but for Paper;
- Typing `git status` should show that we are 10 or 11 commits ahead of master,
meaning we have 10 or 11 patches Spigot and CraftBukkit don't.
   - If it says something like `212 commits ahead, 207 commits behind`,
   cd into the root directory of the cloned repository and type `git fetch` to
   update your upstream. Setting up a remote for the upstream Paper repository
   might be necessary.
```

## Adding Patches

Adding patches to Paper is very simple:

1. Modify `Paper-Server` and/or `Paper-API` with the appropriate changes;
1. Type `git add .` inside these directories to add your changes;
1. Run `git commit` with the desired patch message;
1. Run `./paper rebuild` in the main directory to convert your commit into a new
patch;

> 1. PR your patches back to this repository.

Reviewer: ```suggestion
1. PR the newly generated patch file(s) back to this repository.
```

Your commit will be converted into a patch that you can then PR into Paper.

> ❗ Please note that if you have some specific implementation detail you'd like
> to document, you should do so in the patch message *or* in comments.

Reviewer 3: ```suggestion
> ❗ Please note that if you have some specific implementation detail you'd like
to document, you should do so in the patch message *or* in comments.
```
Placing noteworthy messages in blockquotes will make them more noticeable than just an emoji at the start of the sentence

Author: The block quotes are (sadly) not rendered in GitHub, but I've changed to use them for anyone viewing the document in something like IntelliJ's markdown previewer.

## Modifying Patches

Modifying previous patches is a bit more complex:

### Method 1

This method works by temporarily resetting your `HEAD` to the desired commit to
edit it using `git rebase`.

> ❗ While in the middle of an edit, you will not be able to compile unless you
> *also* reset the opposing module(s) to a related commit. In the API's case, you
> must reset the Server, and reset the API if you're editing the Server.

Reviewer 3: ```suggestion
> ❗ While in the middle of an edit, you will not be able to compile unless you
*also* reset the opposing module(s) to a related commit. In the API's case, you
must reset the Server, and reset the API if you're editing the Server.
```
Placing noteworthy messages in blockquotes will make them more noticeable than just an emoji at the start of the sentence

#### Using the Paper tool

The PaperMC build tool provides a handy command to automatically do this type of
patch modification.

1. Type `./paper edit server` or `./paper edit api` depending on which project
you want to edit;
   - It should show something like
   [this](https://gist.github.com/zachbr/21e92993cb99f62ffd7905d7b02f3159) in
   the text editor you get.
   - If your editor does not have a "menu" at the bottom, you're using `vim`.  
   If you don't know how to use `vim` and don't want to
   learn, enter `:q!` and press enter. Before redoing this step, do
   `export EDITOR=nano` for an easier editor to use.
1. Replace `pick` with `edit` for the commit/patch you want to modify, and
"save" the changes;
   - Only do this for **one** commit at a time.
1. Make the changes you want to make to the patch;
1. Type `./paper edit continue` in the root directory to finish and rebuild
patches;
1. PR your modifications back to this repository.

#### Manual method: Stashing

In case you need something more complex or want more control, this step-by-step instruction does

> exactly what the above slightly automated system does.

Reviewer: ```suggestion
In case you need to do something more complex or just want more control, these step-by-step instructions do exactly what the automated system above does.
```

> 1. If you have changes you are working on type `git stash` to store them for
> later;

Reviewer 3: ```suggestion
1. If you have changes you are working on, type `git stash` to store them for
later;
```
missing comma

   - You can type `git stash pop` to get them back at any point.
1. Type `git rebase -i upstream/upstream`;
   - It should show something like
   [this](https://gist.github.com/zachbr/21e92993cb99f62ffd7905d7b02f3159) in
   the text editor you get.
   - If your editor does not have a "menu" at the bottom, you're using `vim`.  
   If you don't know how to use `vim` and don't want to
   learn, enter `:q!` and press enter. Before redoing this step, do
   `export EDITOR=nano` for an easier editor to use.
1. Replace `pick` with `edit` for the commit/patch you want to modify, and
"save" the changes;
   - Only do this for **one** commit at a time.
1. Make the changes you want to make to the patch;
1. Type `git add .` to add your changes;
1. Type `git commit --amend` to commit;
   - **Make sure to add `--amend`** or else a new patch will be created.
   - You can also modify the commit message and author here.
1. Type `git rebase --continue` to finish rebasing;
1. Type `./paper rebuild` in the root directory;
   - This will modify the appropriate patches based on your commits.

> 1. PR your modifications back to this repositry.

Reviewer 3: ```suggestion
1. PR your modifications back to this repository.
```
spelling mistake

> 1. PR your modifications back to this repository.

Reviewer: ```suggestion
1. PR your modified patch file(s) back to this repository.
```

  message.
   1. `s`/`squash`: Merge your changes into the patch and use your commit message
  and subject.
1. Type `./paper rebuild` in the root directory;
   - This will modify the appropriate patches based on your commits.
1. PR your modifications back to this repository.

#### Automatic method

1. Make your change while at HEAD;
1. Make a fixup commit. `git commit -a --fixup <hashOfPatchToFix>`;
   - You can also use `--squash` instead of `--fixup` if you want the commit
   message to also be changed.
   - You can get the hash by looking at `git log` or `git blame`; your IDE can
  assist you too.
   - Alternatively, if you only know the name of the patch, you can do
  `git commit -a --fixup "Subject of Patch name"`.
1. Rebase with autosquash: `git rebase --autosquash -i upstream/upstream`.
This will automatically move your fixup commit to the right place, and you just
need to "save" the changes.
1. Type `./paper rebuild` in the root directory;
   - This will modify the appropriate patches based on your commits.
1. PR your modifications back to this repository.

## PR Policy

We'll accept changes that make sense. You should be able to justify their

> existance, along with any maintenance costs that come with them. Using

Reviewer 2: existence

> obfuscation helpers helps in the maintenance costs. Remember that these changes

Reviewer 2: helps -> aids

> obfuscation helpers aids in the maintenance costs. Remember that these changes

Reviewer: ```suggestion
[obfuscation helpers](#obfuscation-helpers) aids in the maintenance costs. Remember that these changes
```

will affect everyone who runs Paper, not just you and your server.

While we will fix minor formatting issues, you should stick to the guide below
when making and submitting changes.

## Formatting

All modifications to non-Paper files should be marked.

- Multi line changes start with `// Paper start` and end with `// Paper end`;

> - You can put a messages with a change if it isn't obvious, like this:

Reviewer 3: ```suggestion
- You can put markings with a comment if it isn't obvious, like this:
```
```suggestion
- You can put messages with a change if it isn't obvious, like this:
```
```suggestion
- You can put an explanation after the marker if it isn't obvious, like this:
```
wording suggestions

`// Paper start - reason`.
   - The comments should generally be about the reason the change was made, what
  it was before, or what the change is.
   - Multi-line messages should start with `// Paper start` and use `/* Multi
  line message here */` for the message itself.
- One-line changes should have `// Paper` or `// Paper - reason`.

Here's an example of how to mark changes by Paper:

```java
entity.getWorld().dontbeStupid(); // Paper - was beStupid() which is bad

> entity.getFriends().forEach(Entity::explode);

Reviewer: forEach WeirdChamp

  * [My commit doesn't need a build, what do I do?](#my-commit-doesnt-need-a-build-what-do-i-do)
  * [Patching and building is *really* slow, what can I do?](#patching-and-building-is-really-slow-what-can-i-do)
  * [I wrote some API, how do I use it in Paper-Server?](#i-wrote-some-api-how-do-i-use-it-in-paper-server)

<!-- vim-markdown-toc -->

## Use a Personal Fork and not Organization

Paper will routinely modify your PR, whether it's a quick rebase or to take care
of any minor nitpicks we might have. Often, it's better for us to solve these
problems for you than make you go back and forth trying to fix it yourself.

Unfortunately, if you use an organization for your PR, it prevents Paper from
modifying it. This requires us to manually merge your PR, resulting in us
closing the PR instead of marking it as merged.

We much prefer to have PRs show as merged, so please do not use repositories
on organizations for PRs.

See <https://github.com/isaacs/github/issues/1681> for more information on the
issue.

## Requirements

To get started with PRing changes, you'll need the following software, most of

> which are all available in (most) package managers such as `apt` (Debian /

Reviewer 4: ```suggestion
which is available in package managers such as `apt` (Debian /
```
software is not countable

Author: `most of which is` when talking about a mass noun (according to [Oxford dictionaries](https://www.lexico.com/definition/software) anyway) is not correct either, I believe. I'll wait for someone else to kick in with a correction if I'm wrong.

Reviewer 2: I think you're wrong. "software, most of which are available" sounds off-putting. Software is, not "software are".

Author: I just reworded it as a whole to avoid potential pluralisation.

Reviewer 4: "Mass-noun" is just a fancy term for uncountable.. I also don't see anything on that page that suggests you'd treat it as a normal plural (with "are") 👀

Ubuntu), `homebrew` (macOS / Linux), and more:

- `git` (package `git` everywhere);
- `patch` (often package `patch`);
- A Java 8 or later JDK (packages vary, use Google/DuckDuckGo/etc.).  
If you need one, you can find them on [AdoptOpenJDK](https://adoptopenjdk.net/).
- `maven` (often package `maven`; can be found on
[Apache's site](https://maven.apache.org/download.cgi) too).

If you're on Windows, check
[the section on WSL](#patching-and-building-is-really-slow-what-can-i-do).

If you're compiling with Docker, you can use the
[`adoptopenjdk`](https://hub.docker.com/_/adoptopenjdk/) images like so:

```console
# docker run -it -v "$(pwd)":/data --rm adoptopenjdk:8-jdk-hotspot bash
Pulling image...

root@abcdefg1234:/# javac -version
javac 1.8.0_252
```

## Understanding Patches

Paper is essentially a set of patches on top of Spigot. These patches are split
into different directories which target a certain part of the code. These
directories are:

- `Paper-API` - Modifications to `Spigot-API`/`Bukkit`;

> - `Paper-MojangAPI` - An API for

Reviewer 4: "Paper is essentially a set of patches on top of Spigot" - MojangAPI doesn't really fit under that description (so either change the introduction above or move the MojangAPI separated under the patch directories)

[Mojang's Brigadier](https://github.com/Mojang/brigadier);
- `Paper-Server` - Modifications to `Spigot`/`CraftBukkit`.

Because the entire structure is based on patches and git, a basic understanding
of how to use git is required. A basic tutorial can be found here:
<https://git-scm.com/docs/gittutorial>.

> Assuming you already have forked the repository:

Reviewer 4: ```suggestion
Assuming you have already forked the repository:
```
(most adverbs of time go into mid position)

> you must reset the Server, and reset the API if you're editing the Server.

#### Using the Paper tool

The PaperMC build tool provides a handy command to automatically do this type of
patch modification.

1. Type `./paper edit server` or `./paper edit api` depending on which project
you want to edit;
   - It should show something like
   [this](https://gist.github.com/zachbr/21e92993cb99f62ffd7905d7b02f3159) in
   the text editor you get.
   - If your editor does not have a "menu" at the bottom, you're using `vim`.  
   If you don't know how to use `vim` and don't want to
   learn, enter `:q!` and press enter. Before redoing this step, do
   `export EDITOR=nano` for an easier editor to use.
1. Replace `pick` with `edit` for the commit/patch you want to modify, and
"save" the changes;
   - Only do this for **one** commit at a time.
1. Make the changes you want to make to the patch;
1. Type `./paper edit continue` in the root directory to finish and rebuild
patches;
1. PR your modified patch file(s) back to this repository.

#### Manual method: Stashing

In case you need something more complex or want more control, this step-by-step

> instruction do exactly what the above slightly automated system does.

Reviewer 4: ```suggestion
instruction does exactly what the slightly automated system above does.
```

Location spawnLocation = ((CraftWorld)entity.getWorld()).getSpawnLocation();
entity.getWorld().explode(new BlockPosition(spawnLocation.getX(), spawnLocation.getY(), spawnLocation.getZ()));
// Paper end
```

We generally follow usual Java style (aka. Oracle style), or what is programmed
into most IDEs and formatters by default. There are a few notes, however:
- It is fine to go over 80 lines as long as it doesn't hurt readability.  
There are exceptions, especially in Spigot-related files

> - When in doubt or the code around is in an a clearly different style, use the

Reviewer 2: > in an a

-> in a

> - When in doubt or the code around is in an a clearly different style, use the
> same style as the surrounding code.

Reviewer 3: ```suggestion
- When in doubt or the code around your change is in an a clearly different style, use the
same style as the surrounding code.
```
words

```

## Obfuscation Helpers

In an effort to make future updates easier on ourselves, Paper tries to use
obfuscation helpers whenever possible. The purpose of these helpers is to make
the code more readable and maintainable. These helpers should be be made as easy
to inline as possible by the JVM whenever possible.

An obfuscation helper to get an obfuscated field may be as simple as something like this:

```java
public final int getStuckArrows() { return this.bY(); } // Paper - OBFHELPER
```

> Or it may be as complex as forwarding an entire method so that it can be overriden later:

Reviewer 3: ```suggestion
Or it may be as complex as forwarding an entire method so that it can be overridden later:
```
spelling

```

## Obfuscation Helpers

In an effort to make future updates easier on ourselves, Paper tries to use
obfuscation helpers whenever possible. The purpose of these helpers is to make
the code more readable and maintainable. These helpers should be be made as easy
to inline as possible by the JVM whenever possible.

> An obfuscation helper to get an obfuscated field may be as simple as something

Reviewer: ```suggestion
An obfuscation helper to get an obfuscated method may be as simple as something
```

    return true;
}
```

> While they may not always be done in exactly the same way each time, the general

Reviewer 2: Remove "each time"

goal is always to improve readability and maintainability, so use your best
judgement and do what fits best in your situation.

## Configuration files

To use a configurable value in your patch, add a new entry in either the
`PaperConfig` or `PaperWorldConfig` classes. Use the former (`PaperConfig`) if
a value must remain the same throughout all worlds, or the latter if it can
change between worlds. The latter is preferred whenever possible.

### PaperConfig example

```java
public static boolean saveEmptyScoreboardTeams = false;
private static void saveEmptyScoreboardTeams() {
    // This is called automatically!
    // The name also doesn't matter.
    saveEmptyScoreboardTeams = getBoolean("settings.save-empty-scoreboard-teams", false);
}
```

Notice that the field is always public, but the setter is always private. This
is important to the way the configuration generation system works. To access
this value, reference it as you would any other static value:

```java
if (!PaperConfig.saveEmptyScoreboardTeams) {
```

It is often preferred that you use the fully qualified name for the
configuration class when accessing it, like so:
`com.destroystokyo.paper.PaperConfig.valueHere`.  

> If this is preferred, a developer for Paper might just fix that before merging,

Reviewer 2: > If this is preferred,

-> If this is not done,

}
```

Again, notice that the field is always public, but the setter is always private.
To access this value, you'll need an instance of the `net.minecraft.World`
object:

```java
return this.world.paperConfig.useInhabitedTime ? this.w : 0;
```

## Frequently Asked Questions

### I can't find the NMS file I need!

By default, Paper (and upstream) only import files we make changes to. If you
would like to make changes to a file that isn't present in `Paper-Server`'s
source directory, you just need to add it to our import script ran during the
patching process.

1. Save (rebuild) any patches you are in the middle of working on! Their
progress will be lost if you do not;
1. Identify the name(s) of the file(s) you want to import.
   - A complete list of all possible file names can be found at
   `./work/Minecraft/$MCVER/spigot/net/minecraft/server`. You might find
   [MiniMappingViewer] very useful to find the file you need.
1. Open the file at `./scripts/importmcdev.sh` and add the name of your file to
the script. Skip to the 2nd last header and follow the instructions there;
1. Re-patch the server `./paper patch`;
1. Edit away!

> ❗ This change is temporary! **DO NOT COMMIT CHANGES TO THIS FILE!**  

Reviewer 3: ```suggestion
> ❗ This change is temporary! **DO NOT COMMIT CHANGES TO THIS FILE!**  
```
Placing noteworthy messages in blockquotes will make them more noticeable than just an emoji at the start of the sentence

method arguments as well, whereas Mojang's do not. If you need local variables
to understand the code, you might be more lucky with ModCoderPack.

For more information on the Mojang name licensing issues, check this out:
<https://cpw.github.io/MinecraftMappingData.html>

### My commit doesn't need a build, what do I do?

Well, quite simple: You add `[CI-SKIP]` to the start of your commit subject.

This case most often applies to changes to files like `README.md`, this very
file (`CONTRIBUTING.md`), the `LICENSE.md` file, and so forth.

### Patching and building is *really* slow, what can I do?

This only applies if you're running Windows 10. If you're running a prior
Windows release, either update to Windows 10 or move to macOS/Linux/BSD.

To speed up the programs used, you have to get WSL 2. This is available in
Windows 10 v2004, build 19041 or higher. Check your version by running `winver`
in the run window (Windows key + R). If you're out of date, update your system
with the Windows Update Assistant.

To get WSL 2, follow the information at this link:
<https://docs.microsoft.com/en-us/windows/wsl/install-win10>

> You will most likely get most out of using the Ubuntu apps. Once it's set up,
> install the required toolswith `sudo apt-get update && sudo apt-get install
> git patch default-jdk maven -y`. You can now clone the repository and do
> everything like usual.

Reviewer 3: ```suggestion
You will most likely get the most out of it by using the Ubuntu apps. Once it's set up,
install the required tools with `sudo apt-get update && sudo apt-get install
git patch default-jdk maven -y`. You can now clone the repository and do
everything like usual.
```
wording and spacing

> ❗ Do not use the `/mnt/` directory in WSL! Instead, mount the WSL directories
> in Windows like described here:
> <https://www.howtogeek.com/426749/how-to-access-your-linux-wsl-files-in-windows-10/>

Reviewer 3: ```suggestion
> ❗ Do not use the `/mnt/` directory in WSL! Instead, mount the WSL directories
in Windows like described here:
<https://www.howtogeek.com/426749/how-to-access-your-linux-wsl-files-in-windows-10/>
```
Placing noteworthy messages in blockquotes will make them more noticeable than just an emoji at the start of the sentence

    return true;
}
```

While they may not always be done in exactly the same way, the general goal is

> always to improve readability and maintainability, so use your best judgement

Reviewer: ```suggestion
always to improve readability and maintainability. Use your best judgment
```

and do what fits best in your situation.

## Configuration files

To use a configurable value in your patch, add a new entry in either the
`PaperConfig` or `PaperWorldConfig` classes. Use the former (`PaperConfig`) if
a value must remain the same throughout all worlds, or the latter if it can

> change between worlds. The latter is preferred whenever possible.

Reviewer: ```suggestion
To use a configurable value in your patch, add a new entry in either the
`PaperConfig` or `PaperWorldConfig` classes. Use `PaperConfig` if a value
remains the same throughout all worlds, or `PaperWorldConfig` if it can
change between worlds. World-specific config options are preferred whenever
possible.
```

would like to make changes to a file that isn't present in `Paper-Server`'s
source directory, you just need to add it to our import script ran during the
patching process.

1. Save (rebuild) any patches you are in the middle of working on! Their
progress will be lost if you do not;
1. Identify the name(s) of the file(s) you want to import.
   - A complete list of all possible file names can be found at
   `./work/Minecraft/$MCVER/spigot/net/minecraft/server`. You might find
   [MiniMappingViewer] very useful to find the file you need.
1. Open the file at `./scripts/importmcdev.sh` and add the name of your file to
the script. Skip to the 2nd last header and follow the instructions there;
1. Re-patch the server `./paper patch`;
1. Edit away!

> ❗ This change is temporary! **DO NOT COMMIT CHANGES TO THIS FILE!**  
> Once you have made your changes to the new file, and rebuilt patches, you may
> undo your changes to `importmcdev.sh`.

Any file modified in a patch file gets automatically imported, so you only need
this temporarily to import it to create the first patch.

To undo your changes to the file, type `git checkout scripts/importmcdev.sh`.

> ### Where can I learn what this method/field does?

Reviewer: ```suggestion
### Where can I learn how to name this method/field?
```

For most cases, it is preferred if you use [yarn], as their license works with
Paper's license. If you can't do that, [MiniMappingViewer] is always around to
provide you with more of a general idea, *but* you cannot use the Mojang names
for more than understanding the code.

[yarn] is in general more thorough than Mojang's own mappings, as they include
method arguments as well, whereas Mojang's do not. If you need local variables
to understand the code, you might be more lucky with ModCoderPack.

For more information on the Mojang name licensing issues, check this out:
<https://cpw.github.io/MinecraftMappingData.html>

### My commit doesn't need a build, what do I do?

Well, quite simple: You add `[CI-SKIP]` to the start of your commit subject.

This case most often applies to changes to files like `README.md`, this very
file (`CONTRIBUTING.md`), the `LICENSE.md` file, and so forth.

### Patching and building is *really* slow, what can I do?

> This only applies if you're running Windows 10. If you're running a prior
> Windows release, either update to Windows 10 or move to macOS/Linux/BSD.

Reviewer: ```suggestion
This only applies if you're running Windows. If you're running a prior
Windows release, either update to Windows 10 or move to macOS/Linux/BSD.
```

> To speed up the programs used, you have to get WSL 2. This is available in
> Windows 10 v2004, build 19041 or higher. Check your version by running `winver`
> in the run window (Windows key + R). If you're out of date, update your system
> with the Windows Update Assistant.

Reviewer: ```suggestion
In order to speed up the patching process on Windows, it's recommended you get WSL 2.
WSL2 is available in Windows 10 v2004, build 19041 or higher. (You can check your
Windows version by running `winver` in the run window (Windows key + R)). If you're out
of date, update your system with the [Windows Update Assistant](https://www.microsoft.com/en-us/software-download/windows10).
```

> To get WSL 2, follow the information at this link:

Reviewer: ```suggestion
To setup WSL 2, follow the information here:
```
